src/Functions/tokens.cpp

```diff
@@ -0,0 +1,166 @@
+    size_t getNumberOfArguments() const override { return 0; }
+    bool isVariadic() const override { return true; }
+    bool useDefaultImplementationForConstants() const override { return true; }
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        FunctionArgumentDescriptors mandatory_args{
+            {"value", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isStringOrFixedString), nullptr, "String or FixedString"}};
+
+        FunctionArgumentDescriptors optional_args;
+
+        if (arguments.size() > 1)
+        {
+            optional_args.emplace_back("tokenizer", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isString), isColumnConst, "String");
+
+            validateFunctionArguments(*this, {arguments[arg_value], arguments[arg_tokenizer]}, mandatory_args, optional_args);
+
+            if (arguments.size() == 3)
+            {
+                const auto tokenizer = arguments[arg_tokenizer].column->getDataAt(0).toString();
+
+                if (tokenizer == NgramTokenExtractor::getExternalName())
+                    optional_args.emplace_back("ngrams", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isUInt8), isColumnConst, "UInt8");
+#if USE_CPPJIEBA
+                else if (tokenizer == ChineseTokenExtractor::getExternalName())
+                    optional_args.emplace_back("chinese_granularity", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isString), isColumnConst, "String");
+#endif
+            }
+        }
+
+        validateFunctionArguments(*this, arguments, mandatory_args, optional_args);
+
+        return std::make_shared<DataTypeArray>(std::make_shared<DataTypeString>());
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+    {
+        auto col_input = arguments[arg_value].column;
+        auto col_result = ColumnString::create();
+        auto col_offsets = ColumnArray::ColumnOffsets::create();
+
+        if (input_rows_count == 0)
+            return ColumnArray::create(std::move(col_result), std::move(col_offsets));
+
+        std::unique_ptr<ITokenExtractor> token_extractor;
+
+        const auto tokenizer_arg = arguments.size() < 2 ? SplitTokenExtractor::getExternalName()
+                                                        : arguments[arg_tokenizer].column->getDataAt(0).toView();
+
+        if (tokenizer_arg == SplitTokenExtractor::getExternalName())
+            token_extractor = std::make_unique<SplitTokenExtractor>();
+        else if (tokenizer_arg == NoOpTokenExtractor::getExternalName())
+            token_extractor = std::make_unique<NoOpTokenExtractor>();
+        else if (tokenizer_arg == NgramTokenExtractor::getExternalName())
+        {
+            auto ngrams = arguments.size() < 3 ? 3
```

Comment:
Minor change: Afaik, the 3rd ngrams argument was mandatory before if `ngrams` was given as tokenizer. Now its optional.